src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -279,38 +282,51 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
             {
                 return null;
             }
-
             var headers = httpContext.Request.Headers;
-            var requestId = headers.TraceParent;
-            if (requestId.Count == 0)
-            {
-                requestId = headers.RequestId;
-            }
+            _propagator.ExtractTraceIdAndState(headers,
+                static (object? carrier, string fieldName, out string? fieldValue, out IEnumerable<string>? fieldValues) =>
+                {
+                    fieldValue = default;
+                    fieldValues = default;
+                    if (carrier is null)
+                    {
+                        return;
+                    }
+                    var headers = (IHeaderDictionary)carrier;
+                    fieldValue = headers[fieldName];
+                },
+                out var requestId,
+                out var traceState);
 
-            if (!StringValues.IsNullOrEmpty(requestId))
+            if (requestId is not null)
             {
                 activity.SetParentId(requestId);
-                var traceState = headers.TraceState;
-                if (traceState.Count > 0)
-                {
-                    activity.TraceStateString = traceState;
-                }
-
-                // We expect baggage to be empty by default
-                // Only very advanced users will be using it in near future, we encourage them to keep baggage small (few items)
-                var baggage = headers.GetCommaSeparatedValues(HeaderNames.Baggage);
-                if (baggage.Length == 0)
+            }
+            if (traceState is not null)
+            {
+                activity.TraceStateString = traceState;
+            }
+            
+            if (!string.IsNullOrEmpty(requestId))
```

Comment:
>if (!string.IsNullOrEmpty(requestId)) [](http://example.com/codeflow?start=12&length=37)

can we move this block inside previous block checking on request Id?

Reply:
(y)